### ISSUE TITLE
EditorConfig file in Controllers folder causes COMException (0x80004005)

### ISSUE THREAD

**Issue by ImrePyhvel** [T4MVC/T4MVC#124]
Using VS2019 (16.4.3).

**Repro steps:**
1. Open any MVC project or create a new one with at least 1 controller in Controllers folder
2. Add empty .editorconfig file to Controllers folder
3. t4MVC.tt -> "Run custom tool"

**Expected result:**
* all controllers are handled as before
* .editorConfig file is ignored as not a controller

**Actual result:**
Code generation interrupted with ugly non-descriptive exception:

> Running transformation: System.Runtime.InteropServices.COMException (0x80004005): Error HRESULT E_FAIL has been returned from a call to a COM component.
>    at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
>    at Microsoft.VisualStudio.LanguageServices.Implementation.Utilities.Exceptions.ThrowEFail()
>    at Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.FileCodeModel.GetDocument()
>    at Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.FileCodeModel.GetSyntaxTree()
>    at Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.FileCodeModel..ctor(CodeModelState state, Object parent, DocumentId documentId, ITextManagerAdapter textManagerAdapter)
>    at Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.CodeModelProjectCache.GetOrCreateFileCodeModel(String filePath, Object parent)
>    at Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.ProjectCodeModel.Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.IProjectCodeModel.GetOrCreateFileCodeModel(String filePath, Object parent)
>    at Microsoft.VisualStudio.LanguageServices.CSharp.ProjectSystemShim.CSharpProjectShim.CreateFileCodeModel(String fileName, Object parent, FileCodeModel& ppFileCodeModel)
>    at EnvDTE.ProjectItem.get_FileCodeModel()
>    at Microsoft.VisualStudio.TextTemplatingBA3C17665E04F379CB39E59EFBAFF58DBE3B35B599710FDD2D72852C5182ABC47917AB9CD6A18A81D1AA4EBFC739BA14D3AFA6C882E4ED225D007A352DEA485C.GeneratedTextTransformation.ProcessControllersRecursive(ProjectItem projectItem, AreaInfo area) in C:\Code\WebApplication1\WebApplication1\T4MVC.tt:line 725
>    at Microsoft.VisualStudio.TextTemplatingBA3C17665E04F379CB39E59EFBAFF58DBE3B35B599710FDD2D72852C5182ABC47917AB9CD6A18A81D1AA4EBFC739BA14D3AFA6C882E4ED225D007A352DEA485C.GeneratedTextTransformation.ProcessControllersRecursive(ProjectItem projectItem, AreaInfo area) in C:\Code\WebApplication1\WebApplication1\T4MVC.tt:line 722
>    at Microsoft.VisualStudio.TextTemplatingBA3C17665E04F379CB39E59EFBAFF58DBE3B35B599710FDD2D72852C5182ABC47917AB9CD6A18A81D1AA4EBFC739BA14D3AFA6C882E4ED225D007A352DEA485C.GeneratedTextTransformation.ProcessAreaControllers(ProjectItems areaFolderItems, AreaInfo area, ProjectItem areaFolder) in C:\Code\WebApplication1\WebApplication1\T4MVC.tt:line 692
>    at Microsoft.VisualStudio.TextTemplatingBA3C17665E04F379CB39E59EFBAFF58DBE3B35B599710FDD2D72852C5182ABC47917AB9CD6A18A81D1AA4EBFC739BA14D3AFA6C882E4ED225D007A352DEA485C.GeneratedTextTransformation.ProcessProjectRoot(Project project) in C:\Code\WebApplication1\WebApplication1\T4MVC.tt:line 615
>    at Microsoft.VisualStudio.TextTemplatingBA3C17665E04F379CB39E59EFBAFF58DBE3B35B599710FDD2D72852C5182ABC47917AB9CD6A18A81D1AA4EBFC739BA14D3AFA6C882E4ED225D007A352DEA485C.GeneratedTextTransformation.PrepareDataToRender(TextTransformation tt) in C:\Code\WebApplication1\WebApplication1\T4MVC.tt:line 591
>    at Microsoft.VisualStudio.TextTemplatingBA3C17665E04F379CB39E59EFBAFF58DBE3B35B599710FDD2D72852C5182ABC47917AB9CD6A18A81D1AA4EBFC739BA14D3AFA6C882E4ED225D007A352DEA485C.GeneratedTextTransformation.TransformText() in C:\Code\WebApplication1\WebApplication1\T4MVC.tt:line 40	WebApplication1	C:\Code\WebApplication1\WebApplication1\T4MVC.tt	725

**Comment by davidebbo** [comment]
I'm no longer active in the .NET world, but I would suggest trying to isolate. It may come down to a VS bug. Here, t4mvc is just accessing the `FileCodeModel` property on a ProjectItem:

```c#
    if (projectItem.FileCodeModel != null) { ... }
```

And that's blowing up cryptically somewhere is VS's `CodeModel` logic, which doesn't seem normal. It would be interesting to find out what specific ProjectItem it's happening on. Try dumping `projectItem.get_FileNames(0)` just before that call.

This doesn't happen when you don't have the .editorConfig? It may be possible to work around on the T4MVC side by special casing that file.

**Comment by ImrePyhvel** [comment]
I did dump `projectItem.get_FileNames(0)`, that's how I got the pointer to `.editorconfig` file.

I have only experienced this with editorconfig file. File contents does not seem to matter,  I got repro also with an empty file. 

Interestingly files like `a.editorconfig` nor `.editorconfig2` do not trigger the error. So it seems to be some conflict with exactly `.editorconfig`.  

### Workaround 1
One can always refactor EditorConfig rules to allow the file to reside somewhere further up the folder chain. 

### Workaround 2
Modify the tt to exclude processing the `.editorconfig` file (and remembering to reapply the hack on every upgrade). 

## Possible fix
Considering the FileCodeModel property behavior seems to be out-of-scope for T4MVC, then I see 3 ways to tackle this:
1. Track the owner of FileCodeModel and try to get it to get an understanding what this error actually means and possible get a fix from their side. It could be the correct way to go, but it is likely to be a longer journey. 
1. Add hardcoded exclusion for `.editorconfig` to T4MVC.tt. It is simplest, but it may hide some deeper issue and may not cover other variants of the same issue, triggered by some other files.
1. Add try-catch to FileCodeModel access to handle  the exception. May get a more universal fix, but it is tricky: what exactly to catch? How exactly should the exception be handled? 

Thoughts?

**Comment by davidebbo** [comment]
Thanks for your analysis. It describes our options very well!

I think 2) is the best option to unblock things. 3) may end up masking some unrelated issues, and until we see signs that this is a more general problem, it's seems uncalled for.

Ideally, we'd *also* do 1), as there is likely a bug there. The hard part is to track this down. It was hard enough when I worked at Microsoft. :) Usually, the first step to reporting such issue is to create a super minimal repro, so they don't get confused by giant t4mvc. i.e. if you can write a < 10 line .tt file running in a project that mostly just has the `.editorconfig` and that makes it blow up, maybe there is a chance someone will pick it up.

**Comment by davidebbo** [comment]
Ok, I put in the workaround in 4.2.3. 32948860ab0cb3f747c5be035922e348ba6fa8ca

**Comment by davidebbo** [comment]
I don't have time to track the root issue with the VS team, and my guess is that it would be a lot of effort that may not lead anywhere. But feel free if you want to take that on :)

Looks like this is a related repo: https://github.com/editorconfig/editorconfig-visualstudio
